api/v3/PaymentSchedule.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+use CRM_MembershipExtras_ExtensionUtil as E;
+
+function _civicrm_api3_payment_schedule_get_spec(&$spec) {
+  $spec['membership_type_id'] = [
+    'name' => 'membership_type_id',
+    'title' => 'Membership Type ID',
+    'description' => 'Membership Type ID',
+    'type' => CRM_Utils_Type::T_INT,
+    'FKClassName' => 'CRM_Member_DAO_MembershipType',
+    'FKApiName' => 'MembershipType',
+    'api.required' => 1,
+  ];
+
+  $spec['schedule'] = [
+    'name' => 'schedule',
+    'title' => 'Schedule (monthly, quarterly, annually)',
+    'type' => CRM_Utils_Type::T_STRING,
+    'api.required' => 1,
+  ];
+
+  $spec['start_date'] = [
+    'name' => 'start_date',
+    'title' => 'Membership Start Date',
+    'type' => CRM_Utils_Type::T_DATE,
+    'api.required' => 0,
+  ];
+
+  $spec['end_date'] = [
+    'name' => 'end_date',
+    'title' => 'Membership End Date',
+    'type' => CRM_Utils_Type::T_DATE,
+    'api.required' => 0,
+  ];
+
+  $spec['join_date'] = [
+    'name' => 'join_date',
+    'title' => 'Member Since Date',
+    'type' => CRM_Utils_Type::T_DATE,
+    'api.required' => 0,
+  ];
+}
+
+/**
+ * PaymentScheduleCalculation.get API
+ *
+ * @param array $params
+ * @return array API result descriptor
+ * @throws API_Exception
+ * @throws CRM_MembershipExtras_Exception_InvalidMembershipTypeInstalmentAmount|CiviCRM_API3_Exception
+ * @throws Exception
+ */
+function civicrm_api3_payment_schedule_get($params) {
```

Comment:
@omarabuhussein  I don't mind changing the API name, and the service class. However, on the point 2, I totally agree that we may need to come up with better structure for the extension. However, on the current implementation of this API. The code after line 68 is technically belonging to the API, what it does is to calculate dates for each membership type MembershipTypeDatesCalculator then call MembershipInstalmentsSchedule to generate instalment and finally we format arrays and objects return from generate method and pass the result to civicrm API. 

And for your question for reusing it again in another part inside the extension, we would need to call MembershipInstalmentsSchedule instead of the API.